File: frontend/src/services/workflowTriggers.ts
import { apiRequest, getApiUrl } from "@/config/api";
import {
  WorkflowTrigger,
  WorkflowTriggerDelivery,
  WorkflowTriggerProvisionPayload,
  WorkflowTriggerProvisionResponse,
  WorkflowTriggerRun,
  WorkflowTriggerRunStatus,
  WorkflowTriggerRunSummary,
  WorkflowTriggerSecret,
  WorkflowTriggerTestPayload,
  WorkflowTriggerUpdatePayload,
} from "@/interfaces/workflow-trigger.interface";

const BASE = "genagent/workflow-triggers";

/** Create the endpoint for a trigger node, or fetch the existing one (idempotent). */
export const provisionWorkflowTrigger = async (
  payload: WorkflowTriggerProvisionPayload
): Promise<WorkflowTriggerProvisionResponse> => {
  const base_url = payload.base_url ?? (await getApiUrl());
  const response = await apiRequest<WorkflowTriggerProvisionResponse>(
    "POST",
    `${BASE}/provision`,
    { ...payload, base_url } as unknown as Record<string, unknown>
  );
  if (!response) throw new Error("Failed to provision the webhook endpoint");
  return response;
};

export const getWorkflowTriggersByAgent = async (
  agentId: string
): Promise<WorkflowTrigger[]> => {
  const data = await apiRequest<WorkflowTrigger[]>("GET", `${BASE}/by-agent/${agentId}`);
  return data || [];
};

export const getWorkflowTrigger = async (
  triggerId: string
): Promise<WorkflowTrigger | null> => {
  const data = await apiRequest<WorkflowTrigger>("GET", `${BASE}/${triggerId}`);
  return data ?? null;
};

export const updateWorkflowTrigger = async (
  triggerId: string,
  payload: WorkflowTriggerUpdatePayload
): Promise<WorkflowTrigger> => {
  const response = await apiRequest<WorkflowTrigger>(
    "PUT",
    `${BASE}/${triggerId}`,
    payload as unknown as Record<string, unknown>
  );
  if (!response) throw new Error("Failed to update the webhook endpoint");
  return response;
};

export const deleteWorkflowTrigger = async (triggerId: string): Promise<void> => {
  await apiRequest("DELETE", `${BASE}/${triggerId}`);
};

export const rotateWorkflowTriggerSecret = async (
  triggerId: string
): Promise<WorkflowTriggerSecret> => {
  const response = await apiRequest<WorkflowTriggerSecret>(
    "POST",
    `${BASE}/${triggerId}/rotate-secret`
  );
  if (!response) throw new Error("Failed to rotate the secret");
  return response;
};

export const revealWorkflowTriggerSecret = async (
  triggerId: string
): Promise<WorkflowTriggerSecret> => {
  const response = await apiRequest<WorkflowTriggerSecret>(
    "GET",
    `${BASE}/${triggerId}/reveal-secret`
  );
  if (!response) throw new Error("Failed to reveal the secret");
  return response;
};

export const getWorkflowTriggerRuns = async (
  triggerId: string,
  options?: { status?: WorkflowTriggerRunStatus; limit?: number; offset?: number }
): Promise<WorkflowTriggerRunSummary[]> => {
  const params = new URLSearchParams();
  if (options?.status) params.set("run_status", options.status);
  if (options?.limit !== undefined) params.set("limit", String(options.limit));
  if (options?.offset !== undefined) params.set("offset", String(options.offset));
  const query = params.toString() ? `?${params.toString()}` : "";
  const data = await apiRequest<WorkflowTriggerRunSummary[]>(
    "GET",
    `${BASE}/${triggerId}/runs${query}`
  );
  return data || [];
};

export const getWorkflowTriggerRun = async (
  triggerId: string,
  runId: string
): Promise<WorkflowTriggerRun | null> => {
  const data = await apiRequest<WorkflowTriggerRun>(
    "GET",
    `${BASE}/${triggerId}/runs/${runId}`
  );
  return data ?? null;
};

/** Queue a real run with a sample delivery (skips auth, rate limit and idempotency). */
export const testWorkflowTrigger = async (
  triggerId: string,
  payload: WorkflowTriggerTestPayload
): Promise<WorkflowTriggerDelivery> => {
  const response = await apiRequest<WorkflowTriggerDelivery>(
    "POST",
    `${BASE}/${triggerId}/test`,
    payload as unknown as Record<string, unknown>
  );
  if (!response) throw new Error("Failed to send the test delivery");
  return response;
};

File: backend/app/api/v1/routes/workflow_triggers.py
"""Management API for Webhook Trigger node endpoints.

The public delivery endpoint itself is ``/webhook/execute/{id}`` (see
``webhook_execute.py``); everything here is authenticated builder tooling.
"""

import logging
from typing import List, Optional
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query, Request, status
from fastapi_injector import Injected

from app.auth.dependencies import auth, permissions
from app.core.exceptions.error_messages import ErrorKey
from app.core.exceptions.exception_classes import AppException
from app.core.permissions.constants import Permissions as P
from app.core.utils.enums.workflow_schedule_enum import WorkflowScheduleRunStatus
from app.schemas.workflow_trigger import (
    WorkflowTriggerDeliveryResponse,
    WorkflowTriggerProvision,
    WorkflowTriggerProvisionResponse,
    WorkflowTriggerRead,
    WorkflowTriggerRunRead,
    WorkflowTriggerRunSummary,
    WorkflowTriggerSecret,
    WorkflowTriggerTestRequest,
    WorkflowTriggerUpdate,
)
from app.services.workflow_trigger import WorkflowTriggerService

logger = logging.getLogger(__name__)

router = APIRouter()


def _http(e: AppException) -> HTTPException:
    if e.error_key == ErrorKey.NOT_FOUND:
        return HTTPException(status_code=404, detail=str(e))
    return HTTPException(status_code=400, detail=str(e))


@router.post(
    "/provision",
    response_model=WorkflowTriggerProvisionResponse,
    dependencies=[Depends(auth), Depends(permissions(P.Workflow.UPDATE))],
)
async def provision_workflow_trigger(
    data: WorkflowTriggerProvision,
    request: Request,
    service: WorkflowTriggerService = Injected(WorkflowTriggerService),
):
    """Create the endpoint for a trigger node, or return the existing one."""
    try:
        return await service.provision(data, request)
    except AppException as e:
        raise _http(e)


@router.get(
    "/by-agent/{agent_id}",
    response_model=List[WorkflowTriggerRead],
    dependencies=[Depends(auth), Depends(permissions(P.Workflow.READ))],
)
async def list_workflow_triggers_by_agent(
    agent_id: UUID,
    service: WorkflowTriggerService = Injected(WorkflowTriggerService),
):
    return await service.list_by_agent(agent_id)


@router.get(
    "/{trigger_id}",
    response_model=WorkflowTriggerRead,
    dependencies=[Depends(auth), Depends(permissions(P.Workflow.READ))],
)
async def get_workflow_trigger(
    trigger_id: UUID,
    service: WorkflowTriggerService = Injected(WorkflowTriggerService),
):
    try:
        return await service.get(trigger_id)
    except AppException as e:
        raise _http(e)


@router.put(
    "/{trigger_id}",
    response_model=WorkflowTriggerRead,
    dependencies=[Depends(auth), Depends(permissions(P.Workflow.UPDATE))],
)
async def update_workflow_trigger(
    trigger_id: UUID,
    data: WorkflowTriggerUpdate,
    service: WorkflowTriggerService = Injected(WorkflowTriggerService),
):
    try:
        return await service.update(trigger_id, data)
    except AppException as e:
        raise _http(e)


@router.delete(
    "/{trigger_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    dependencies=[Depends(auth), Depends(permissions(P.Workflow.DELETE))],
)
async def delete_workflow_trigger(
    trigger_id: UUID,
    service: WorkflowTriggerService = Injected(WorkflowTriggerService),
):
    try:
        await service.delete(trigger_id)
        return None
    except AppException as e:
        raise _http(e)


@router.post(
    "/{trigger_id}/rotate-secret",
    response_model=WorkflowTriggerSecret,
    dependencies=[Depends(auth), Depends(permissions(P.Workflow.UPDATE))],
)
async def rotate_workflow_trigger_secret(
    trigger_id: UUID,
    service: WorkflowTriggerService = Injected(WorkflowTriggerService),
):
    try:
        return await service.rotate_secret(trigger_id)
    except AppException as e:
        raise _http(e)


@router.get(
    "/{trigger_id}/reveal-secret",
    response_model=WorkflowTriggerSecret,
    dependencies=[Depends(auth), Depends(permissions(P.Workflow.UPDATE))],
)
async def reveal_workflow_trigger_secret(
    trigger_id: UUID,
    service: WorkflowTriggerService = Injected(WorkflowTriggerService),
):
    try:
        return await service.reveal_secret(trigger_id)
    except AppException as e:
        raise _http(e)


@router.get(
    "/{trigger_id}/runs",
    response_model=List[WorkflowTriggerRunSummary],
    dependencies=[Depends(auth), Depends(permissions(P.Workflow.READ))],
)
async def list_workflow_trigger_runs(
    trigger_id: UUID,
    run_status: Optional[WorkflowScheduleRunStatus] = Query(None, description="Filter by run status"),
    limit: int = Query(50, ge=1, le=100),
    offset: int = Query(0, ge=0),
    service: WorkflowTriggerService = Injected(WorkflowTriggerService),
):
    try:
        return await service.list_runs(trigger_id, status=run_status, limit=limit, offset=offset)
    except AppException as e:
        raise _http(e)


@router.get(
    "/{trigger_id}/runs/{run_id}",
    response_model=WorkflowTriggerRunRead,
    dependencies=[Depends(auth), Depends(permissions(P.Workflow.READ))],
)
async def get_workflow_trigger_run(
    trigger_id: UUID,
    run_id: UUID,
    service: WorkflowTriggerService = Injected(WorkflowTriggerService),
):
    try:
        return await service.get_run(trigger_id, run_id)
    except AppException as e:
        raise _http(e)


@router.post(
    "/{trigger_id}/test",
    response_model=WorkflowTriggerDeliveryResponse,
    status_code=status.HTTP_202_ACCEPTED,
    dependencies=[Depends(auth), Depends(permissions(P.Workflow.EXECUTE))],
)
async def test_workflow_trigger(
    trigger_id: UUID,
    data: WorkflowTriggerTestRequest,
    request: Request,
    service: WorkflowTriggerService = Injected(WorkflowTriggerService),
):
    """Queue a run with a sample delivery, skipping auth and rate limiting."""
    try:
        return await service.test(trigger_id, data, request)
    except AppException as e:
        raise _http(e)
